README
= Description
  A series of extra constants for the Dir class that define special folders
  on MS Windows systems, as well as methods for creating and detecting
  junctions, i.e. symlinks for directories.

= Installation
  gem install win32-dir

= Synopsis
  require 'win32/dir'
	
  # C:\WINNT or C:\WINDOWS
  puts Dir::WINDOWS

  # C:\Documents and Settings\Daniel\Start Menu\Programs\Administrative Tools	 
  puts Dir::ADMINTOOLS
	
  Dir.mkdir('C:\from')
  Dir.create_junction('C:\to', 'C:\from')

= Constants
  Not all of these are guaranteed to be defined on your system. Also note
  that the directories are merely defined. It doesn't necessarily mean they
  actually exist.
	
== The following constants should be defined:	
Dir::ADMINTOOLS
  The file system directory that is used to store administrative tools for an
  individual user. The Microsoft Management Console (MMC) will save
  customized consoles to this directory, and it will roam with the user.

Dir::COMMON_ADMINTOOLS
  The file system directory containing administrative tools for all users
  of the computer.
	
Dir::APPDATA
  The file system directory that serves as a common repository for
  application-specific data. A typical path is
  C:\Documents and Settings\<user>\Application Data.
   
  This CSIDL is supported by the redistributable shfolder.dll for
  systems that do not have the Microsoft Internet Explorer 4.0
  integrated Shell installed.
	
Dir::COMMON_APPDATA
  The file system directory containing application data for all users. A
  typical path is C:\Documents and Settings\All Users\Application Data.
	
Dir::COMMON_DOCUMENTS
  The file system directory that contains documents that are common to all
  users. A typical paths is C:\Documents and Settings\All Users\Documents.
	
Dir::COOKIES
  The file system directory that serves as a common repository for Internet
  cookies. A typical path is C:\Documents and Settings\<user>\Cookies.

Dir::HISTORY
  The file system directory that serves as a common repository for Internet
  history items.

Dir::INTERNET_CACHE
  The file system directory that serves as a common repository for temporary
  Internet files. A typical path is
  C:\Documents and Settings\<user>\Local Settings\Temporary Internet Files.

Dir::LOCAL_APPDATA
  The file system directory that serves as a data repository for local
  (nonroaming) applications. A typical path is
  C:\Documents and Settings\<user>\Local Settings\Application Data.

Dir::MYPICTURES
  The file system directory that serves as a common repository for image
  files. A typical path is
  C:\Documents and Settings\<user>\My Documents\My Pictures.
	
Dir::PERSONAL
  The virtual folder representing the My Documents desktop item. This is
  equivalent to Dir::MYDOCUMENTS.

Dir::PROGRAM_FILES
  The Program Files folder. A typical path is C:\Program Files.
	
Dir::PROGRAM_FILES_COMMON
  A folder for components that are shared across applications. A typical path
  is C:\Program Files\Common.
	
Dir::SYSTEM
  The Windows System folder. A typical path is C:\Windows\System32.
	
Dir::WINDOWS
  The Windows directory or SYSROOT. This corresponds to the %windir% or
  %SYSTEMROOT% environment variables. A typical path is C:\Windows.
	 
== The following constants may or may not be defined:
Dir::ALTSTARTUP
  The file system directory that corresponds to the user's nonlocalized
  Startup program group.
	
Dir::BITBUCKET
  The virtual folder containing the objects in the user's Recycle Bin.
	
Dir::CDBURN_AREA
  The file system directory acting as a staging area for files waiting to
  be written to CD.

Dir::COMMON_ALTSTARTUP
  The file system directory that corresponds to the nonlocalized Startup
  program group for all users.

Dir::COMMON_DESKTOPDIRECTORY
  The file system directory that contains files and folders that appear on
  the desktop for all users. A typical path is
  C:\Documents and Settings\All Users\Desktop.

Dir::COMMON_FAVORITES
  The file system directory that serves as a common repository for favorite
  items common to all users.

Dir::COMMON_MUSIC
  The file system directory that serves as a repository for music files
  common to all users.

Dir::COMMON_PICTURES
  The file system directory that serves as a repository for image files
  common to all users.

Dir::COMMON_PROGRAMS
  The file system directory that contains the directories for the common
  program groups that appear on the Start menu for all users. 

Dir::COMMON_STARTMENU
  The file system directory that contains the programs and folders that
  appear on the Start menu for all users.

Dir::COMMON_STARTUP
  The file system directory that contains the programs that appear in the
  Startup folder for all users.

Dir::COMMON_TEMPLATES
  The file system directory that contains the templates that are available
  to all users.

Dir::COMMON_VIDEO
  The file system directory that serves as a repository for video files
  common to all users.

Dir::CONTROLS
  The virtual folder containing icons for the Control Panel applications.

Dir::DESKTOP
  The virtual folder representing the Windows desktop, the root of the
  namespace.

Dir::DESKTOPDIRECTORY
  The file system directory used to physically store file objects on the
  desktop (not to be confused with the desktop folder itself).

Dir::DRIVES
  The virtual folder representing My Computer, containing everything on
  the local computer: storage devices, printers, and Control Panel. The
  folder may also contain mapped network drives.

Dir::FAVORITES
  The file system directory that serves as a common repository for the
  user's favorite items.

Dir::FONTS
  A virtual folder containing fonts.
	
Dir::INTERNET
  A virtual folder representing the Internet.

Dir::MYDOCUMENTS
  The virtual folder representing the My Documents desktop item.  See also
  Dir::PERSONAL.

Dir::MYMUSIC
  The file system directory that serves as a common repository for music files.

Dir::MYVIDEO
  The file system directory that serves as a common repository for video files.

Dir::NETHOOD
  A file system directory containing the link objects that may exist in the
  My Network Places virtual folder. It is not the same as Dir::NETWORK, which
  represents the network namespace root.

Dir::NETWORK
  A virtual folder representing Network Neighborhood, the root of the network
  namespace hierarchy.

Dir::PRINTERS
  The virtual folder containing installed printers.

Dir::PRINTHOOD
  The file system directory that contains the link objects that can exist in
  the "Printers" virtual folder.

Dir::PROFILE
  The user's profile folder.

Dir::PROFILES
  The file system directory containing user profile folders.

Dir::PROGRAMS
  The file system directory that contains the user's program groups (which
  are themselves file system directories).

Dir::RECENT
  The file system directory that contains shortcuts to the user's most
  recently used documents.

Dir::SENDTO
  The file system directory that contains Send To menu items.

Dir::STARTMENU
  The file system directory containing Start menu items.

Dir::STARTUP
  The file system directory that corresponds to the user's Startup program
  group.

Dir::TEMPLATES
  The file system directory that serves as a common repository for document
  templates.

== Developer's Notes
  The SHGetFolderPath() documentation on MSDN is somewhat vague about which
  CSIDL constants are guaranteed to be defined.  However, there are 15 which
  *should* be defined (see docs above). The rest I cannot vouch for.
   
  Some of these folders are virtual, and the value will be the display name
  only instead of an actual path.
	
== Known Bugs
  The Dir.create_junction and Dir.read_junction methods do not work with JRuby.

  Please log any bug reports on the project page at
  http://www.github.com/djberg96/win32-dir

== Future Plans
  Suggestions welcome.

== Acknowledgements
  Shashank Date and Zach Dennis for the suggestion and supporting comments
  on the mailing list.
	
  Timothy Byrd and Autrijus Tang for help (directly or indirectly) with the
  junction methods. Timothy provided a pure Ruby version of the junction
  code that I later borrowed from.
	
  Most of the documentation was copied from the MSDN web site.

== License
  Artistic 2.0

== Contributions
  Although this library is free, please consider having your company
  setup a gittip if used by your company professionally.

  http://www.gittip.com/djberg96/
    
== Copyright
  (C) 2003-2015 Daniel J. Berger, All Rights Reserved
    
== Warranty
  This package is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.
    
== Authors
  Daniel J. Berger
  Park Heesob


CHANGES
== 0.5.1 - 20-Oct-2015
* The gem is now signed.
* The Rakefile gem tasks now assume Rubygems 2.x.
* Added a win32-dir.rb file for convenience.

== 0.5.0 - 15-Oct-2014
* Fixed an encoding error in an error message in the Dir.read_junction method.
  Thanks go to topac for the spot and patch.
* Replaced File.exists? with File.exist? in the test suite to avoid warnings
  in Ruby 2.1 and later.
* Do not let the version number fool you, this is not a major release. I
  simply ran out of version numbers.

== 0.4.9 - 2-Jul-2014
* Fixed the INVALID_FILE_ATTRIBUTES constant. It's only a 32-bit value.
  Thanks go to Ethan Brown for the patch.
* Use relative_require instead of manually doing things.
* Removed rubyforge_project from gemspec.
* Added a string_check to the Dir.junction? method.

== 0.4.8 - 27-Apr-2014
* More closely mimic Ruby's stringiness checks for overrided methods.

== 0.4.7 - 26-Apr-2014
* All arguments to methods are now interpolated so that they'll use whatever 
  to_str implemenation the argument provides instead of assuming String
  arguments. This change was made because it turns out that at least some
  of the overridden methods in MRI accept Pathname objects. Thanks go to
  Josh Cooper/Puppet Labs for the spot.
* Refactored the Dir.getwd method to use bonafide FFI buffers instead of
  naked Ruby char buffers.
* Internally various functions now fail in the event that the resulting
  paths would exceed their buffers.

== 0.4.6 - 21-Oct-2013
* Fixed the INVALID_HANDLE_VALUE and INVALID_FILE_ATTRIBUTES constants for
  64-bit versions of Ruby.
* Added Rake as a development dependency.

== 0.4.5 - 23-Sep-2013
* Yet another encoding fix, this time one that affected JRuby.
* Fixed the Dir[] and Dir.glob methods so they match the spec. Thanks go
  to Chris Westbrook for the spot.

== 0.4.4 - 22-Sep-2013
* Yet another encoding fix. Thanks go to Rob Reynolds for the patch.

== 0.4.3 - 24-Jul-2013
* Changed the Dir.read_junction and CSIDL constant strings so that they
  use the default external encoding instead of UTF-16LE. This means that
  they will now work with methods like File.join. Thanks go to Josh Cooper
  for the patches.
* Some changes for internal handling of making FFI functions private.
* Updated gem:create task for rubygems 2.

== 0.4.2 - 8-Apr-2013
* Fixed the HANDLE function prototypes in the underlying FFI code and added
  some custom typedefs for convenience. This affects 64 bit Ruby code.

== 0.4.1 - 2-Oct-2012
* Added the Dir.read_junction method. Thanks go to Gabriel Wilkins for the patch.

== 0.4.0 - 29-Jun-2012
* Conversion to FFI. Should work with JRuby now.
* If current versions of Dir::XXX constant values cannot be found
  then default values are tried.
* Now requires Ruby 1.9 or later.

== 0.3.7 - 18-Jul-2010
* Modified Dir.glob and Dir[] to handle backslashes in path names.
* Added tests for the modified Dir.glob and Dir[] behavior.
* Removed the old non-gem install Rake task.

== 0.3.6 - 6-Feb-2010
* Bug fixes for Ruby 1.9.x. Thanks go to Kendall Gifford for the spot and
  the patch.
* Minor refactoring of the Dir.create_junction method. This fixed a bug for
  Ruby 1.9.x, but was also a little cleaner in general.
* Some Rakefile task and gemspec updates.

== 0.3.5 - 6-Aug-2009
* Changed the license to Artistic 2.0.
* Updated the gemspec, including the addition of a license attribute and
  test-unit as a development dependency.

== 0.3.4 - 5-May-2009
* Redefined the Dir.getwd (and the Dir.pwd alias) to always return a
  normalized path.
* Some gemspec updates.

== 0.3.3 - 30-Mar-2009
* Virtual folders like Dir::CONTROL, which were previously almost always nil,
  are now set to their display name.
* Fixed a bug in the create_junction method.
* Added an 'example' rake task to run the example code.
* Renamed the example program from dir_test.rb to dir_example.rb to prevent
  any potential confusion that it's a genuine test file.

== 0.3.2 - 25-Jul-2007
* Added a Rakefile with tasks for testing and installation.
* Removed the install.rb file (the Rakefile handles installation).
* Minor updates the README, MANIFEST, and test file.

== 0.3.1 - 16-Oct-2006
* Added the Dir.empty? method.
* Changed the Dir.reparse_dir? method to Dir.junction? (but kept an alias
  for backwards compatibility).
* Added rdoc for Dir.junction? (oops).
* Some test tweaks and additions.

== 0.3.0 - 28-May-2006
* Now pure Ruby.
* Now has a gem.
* Added a VERSION constant.

== 0.2.0 - 27-Jun-2005
* Added the Dir.create_junction and Dir.reparse_dir? methods.
* Added corresponding tests and documentation.

== 0.1.0 - 25-Feb-2005
* Initial release


examples/dir_example.rb
####################################################################
# dir_example.rb
#
# Generic test script for general futzing.  Modify as you see fit.
# You can run this via the 'rake example' task.
####################################################################
require 'win32/dir'

puts "Admin Tools:\t\t" + Dir::ADMINTOOLS
puts "Common Admin Tools:\t" + Dir::COMMON_ADMINTOOLS
puts "App Data:\t\t" + Dir::APPDATA
puts "Common App Data:\t" + Dir::COMMON_APPDATA
puts "Common Documents:\t" + Dir::COMMON_DOCUMENTS
puts "Cookies:\t\t" + Dir::COOKIES
puts "History:\t\t" + Dir::HISTORY
puts "Internet Cache:\t\t" + Dir::INTERNET_CACHE
puts "Local App Data:\t\t" + Dir::LOCAL_APPDATA
puts "My Pictures:\t\t" + Dir::MYPICTURES
puts "Personal:\t\t" + Dir::PERSONAL
puts "Program Files:\t\t" + Dir::PROGRAM_FILES
puts "Program Files Common:\t" + Dir::PROGRAM_FILES_COMMON
puts "System:\t\t\t" + Dir::SYSTEM
puts "Windows:\t\t" + Dir::WINDOWS 

lib/win32/dir/structs.rb
require 'ffi'

module Dir::Structs
  extend FFI::Library

  class SHFILEINFO < FFI::Struct
    layout(
      :hIcon, :ulong,
      :iIcon, :int,
      :dwAttributes, :ulong,
      :szDisplayName, [:char, 256],
      :szTypeName, [:char, 80]
    )
  end

  # I fudge a bit, assuming a MountPointReparseBuffer
  class REPARSE_JDATA_BUFFER < FFI::Struct
    layout(
      :ReparseTag, :ulong,
      :ReparseDataLength, :ushort,
      :Reserved, :ushort,
      :SubstituteNameOffset, :ushort,
      :SubstituteNameLength, :ushort,
      :PrintNameOffset, :ushort,
      :PrintNameLength, :ushort,
      :PathBuffer, [:char, 1024]
    )

    # The REPARSE_DATA_BUFFER_HEADER_SIZE which is calculated as:
    #
    # sizeof(ReparseTag) + sizeof(ReparseDataLength) + sizeof(Reserved)
    #
    def header_size
      FFI::Type::ULONG.size + FFI::Type::USHORT.size + FFI::Type::USHORT.size
    end
  end
end


lib/win32/dir/functions.rb
# Necessary to force JRuby to use the gem, not its builtin version
if RUBY_PLATFORM == 'java'
  require 'rubygems'
  gem 'ffi'
end

require 'ffi'

module Dir::Functions
  module FFI::Library
    # Wrapper method for attach_function + private
    def attach_pfunc(*args)
      attach_function(*args)
      private args[0]
    end
  end

  extend FFI::Library

  typedef :ulong, :dword
  typedef :uintptr_t, :handle
  typedef :uintptr_t, :hwnd
  typedef :pointer, :ptr

  ffi_lib :shell32

  attach_pfunc :SHGetFolderPathW, [:hwnd, :int, :handle, :dword, :buffer_out], :dword
  attach_pfunc :SHGetFolderLocation, [:hwnd, :int, :handle, :dword, :ptr], :dword
  attach_pfunc :SHGetFileInfo, [:dword, :dword, :ptr, :uint, :uint], :dword

  ffi_lib :shlwapi

  attach_pfunc :PathIsDirectoryEmptyW, [:buffer_in], :bool

  ffi_lib :kernel32

  attach_pfunc :CloseHandle, [:handle], :bool
  attach_pfunc :CreateDirectoryW, [:buffer_in, :ptr], :bool
  attach_pfunc :CreateFileW, [:buffer_in, :dword, :dword, :ptr, :dword, :dword, :handle], :handle
  attach_pfunc :DeviceIoControl, [:handle, :dword, :ptr, :dword, :ptr, :dword, :ptr, :ptr], :bool
  attach_pfunc :GetCurrentDirectoryW, [:dword, :buffer_out], :dword
  attach_pfunc :GetFileAttributesW, [:buffer_in], :dword
  attach_pfunc :GetLastError, [], :dword
  attach_pfunc :GetShortPathNameW, [:buffer_in, :buffer_out, :dword], :dword
  attach_pfunc :GetLongPathNameW, [:buffer_in, :buffer_out, :dword], :dword
  attach_pfunc :GetFullPathNameW, [:buffer_in, :dword, :buffer_out, :ptr], :dword
  attach_pfunc :RemoveDirectoryW, [:buffer_in], :bool
end

class String
  # Convenience method for converting strings to UTF-16LE for wide character
  # functions that require it.
  def wincode
    (self.tr(File::SEPARATOR, File::ALT_SEPARATOR) + 0.chr).encode('UTF-16LE')
  end

  # Read a wide character string up until the first double null, and delete
  # any remaining null characters.
  def wstrip
    self.force_encoding('UTF-16LE').encode('UTF-8',:invalid=>:replace,:undef=>:replace).
    split("\x00")[0].encode(Encoding.default_external)
  end
end


lib/win32/dir.rb
require_relative 'dir/constants'
require_relative 'dir/functions'
require_relative 'dir/structs'

class Dir
  include Dir::Structs
  include Dir::Constants
  extend Dir::Functions

  # The version of the win32-dir library.
  VERSION = '0.5.1'

  # CSIDL constants
  csidl = Hash[
    'DESKTOP',                  0x0000,
    'INTERNET',                 0x0001,
    'PROGRAMS',                 0x0002,
    'CONTROLS',                 0x0003,
    'PRINTERS',                 0x0004,
    'PERSONAL',                 0x0005,
    'FAVORITES',                0x0006,
    'STARTUP',                  0x0007,
    'RECENT',                   0x0008,
    'SENDTO',                   0x0009,
    'BITBUCKET',                0x000a,
    'STARTMENU',                0x000b,
    'MYDOCUMENTS',              0x000c,
    'MYMUSIC',                  0x000d,
    'MYVIDEO',                  0x000e,
    'DESKTOPDIRECTORY',         0x0010,
    'DRIVES',                   0x0011,
    'NETWORK',                  0x0012,
    'NETHOOD',                  0x0013,
    'FONTS',                    0x0014,
    'TEMPLATES',                0x0015,
    'COMMON_STARTMENU',         0x0016,
    'COMMON_PROGRAMS',          0X0017,
    'COMMON_STARTUP',           0x0018,
    'COMMON_FAVORITES',         0x001f,
    'COMMON_DESKTOPDIRECTORY',  0x0019,
    'APPDATA',                  0x001a,
    'PRINTHOOD',                0x001b,
    'LOCAL_APPDATA',            0x001c,
    'ALTSTARTUP',               0x001d,
    'COMMON_ALTSTARTUP',        0x001e,
    'INTERNET_CACHE',           0x0020,
    'COOKIES',                  0x0021,
    'HISTORY',                  0x0022,
    'COMMON_APPDATA',           0x0023,
    'WINDOWS',                  0x0024,
    'SYSTEM',                   0x0025,
    'PROGRAM_FILES',            0x0026,
    'MYPICTURES',               0x0027,
    'PROFILE',                  0x0028,
    'SYSTEMX86',                0x0029,
    'PROGRAM_FILESX86',         0x002a,
    'PROGRAM_FILES_COMMON',     0x002b,
    'PROGRAM_FILES_COMMONX86',  0x002c,
    'COMMON_TEMPLATES',         0x002d,
    'COMMON_DOCUMENTS',         0x002e,
    'CONNECTIONS',              0x0031,
    'COMMON_MUSIC',             0x0035,
    'COMMON_PICTURES',          0x0036,
    'COMMON_VIDEO',             0x0037,
    'RESOURCES',                0x0038,
    'RESOURCES_LOCALIZED',      0x0039,
    'COMMON_OEM_LINKS',         0x003a,
    'CDBURN_AREA',              0x003b,
    'COMMON_ADMINTOOLS',        0x002f,
    'ADMINTOOLS',               0x0030
  ]

  # Dynamically set each of the CSIDL constants
  csidl.each{ |key, value|
    buf  = 0.chr * 1024
    path = nil
    buf.encode!('UTF-16LE')

    if SHGetFolderPathW(0, value, 0, 0, buf) == 0 # Current path
      path = buf.strip
    elsif SHGetFolderPathW(0, value, 0, 1, buf) == 0 # Default path
      path = buf.strip
    else
      ptr   = FFI::MemoryPointer.new(:long)
      info  = SHFILEINFO.new
      flags = SHGFI_DISPLAYNAME | SHGFI_PIDL

      if SHGetFolderLocation(0, value, 0, 0, ptr) == 0
        if SHGetFileInfo(ptr.read_long, 0, info, info.size, flags) != 0
          path = info[:szDisplayName].to_s
          path.force_encoding(Encoding.default_external)
        end
      end
    end

    begin
      Dir.const_set(key, path.encode(Encoding.default_external)) if path
    rescue Encoding::UndefinedConversionError
      Dir.const_set(key, path.encode('UTF-8')) if path
    end
  }

  # Set Dir::MYDOCUMENTS to the same as Dir::PERSONAL if undefined
  unless defined? MYDOCUMENTS
    MYDOCUMENTS = PERSONAL
  end

  class << self
    alias old_glob glob

    # Same as the standard MRI Dir.glob method except that it handles
    # backslashes in path names.
    #
    def glob(glob_pattern, flags = 0, &block)
      if glob_pattern.is_a?(Array)
        temp = glob_pattern.map!{ |pattern| string_check(pattern).tr("\\", "/") }
      else
        temp = string_check(glob_pattern).tr("\\", "/")
      end

      old_glob(temp, flags, &block)
    end

    alias old_ref []

    # Same as the standard MRI Dir[] method except that it handles
    # backslashes in path names.
    #
    def [](*glob_patterns)
      temp = glob_patterns.map!{ |pattern| "#{pattern}".tr("\\", "/") }
      old_ref(*temp)
    end

    # JRuby normalizes the path by default.
    unless RUBY_PLATFORM == 'java'
      alias oldgetwd getwd
      alias oldpwd pwd

      # Returns the present working directory. Unlike MRI, this method always
      # normalizes the path.
      #
      # Examples:
      #
      #    Dir.chdir("C:/Progra~1")
      #    Dir.getwd # => C:\Program Files
      #
      #    Dir.chdir("C:/PROGRAM FILES")
      #    Dir.getwd # => C:\Program Files
      #
      def getwd
        path1 = FFI::Buffer.new(:wint_t, 1024, true)
        path2 = FFI::Buffer.new(:wint_t, 1024, true)
        path3 = FFI::Buffer.new(:wint_t, 1024, true)

        length = GetCurrentDirectoryW(path1.size, path1)

        if length == 0 || length > path1.size
          raise SystemCallError.new("GetCurrentDirectoryW", FFI.errno)
        end

        length = GetShortPathNameW(path1, path2, path2.size)

        if length == 0 || length > path2.size
          raise SystemCallError.new("GetShortPathNameW", FFI.errno)
        end

        length = GetLongPathNameW(path2, path3, path3.size)

        if length == 0 || length > path3.size
          raise SystemCallError.new("GetLongPathNameW", FFI.errno)
        end

        path = path3.read_bytes(length * 2).wstrip

        begin
          path.encode(Encoding.default_external)
        rescue Encoding::UndefinedConversionError
          path.encode('UTF-8')
        end
      end

      alias :pwd :getwd
    end
  end

  # Creates the symlink +to+, linked to the existing directory +from+. If the
  # +to+ directory already exists, it must be empty or an error is raised.
  #
  # Example:
  #
  #    Dir.mkdir('C:/from')
  #    Dir.create_junction('C:/to', 'C:/from')
  #
  def self.create_junction(to, from)
    to   = string_check(to).wincode
    from = string_check(from).wincode

    from_path = (0.chr * 1024).encode('UTF-16LE')

    length = GetFullPathNameW(from, from_path.size, from_path, nil)

    if length == 0 || length > from_path.size
      raise SystemCallError.new("GetFullPathNameW", FFI.errno)
    else
      from_path.strip!
    end

    to_path = (0.chr * 1024).encode('UTF-16LE')

    length = GetFullPathNameW(to, to_path.size, to_path, nil)

    if length == 0 || length > to_path.size
      raise SystemCallError.new("GetFullPathNameW", FFI.errno)
    else
      to_path.strip!
    end

    # You can create a junction to a directory that already exists, so
    # long as it's empty.
    unless CreateDirectoryW(to_path, nil)
      if FFI.errno != ERROR_ALREADY_EXISTS
        raise SystemCallError.new("CreateDirectoryW", FFI.errno)
      end
    end

    begin
      # Generic read & write + open existing + reparse point & backup semantics
      handle = CreateFileW(
        to_path,
        GENERIC_READ | GENERIC_WRITE,
        0,
        nil,
        OPEN_EXISTING,
        FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS,
        0
      )

      if handle == INVALID_HANDLE_VALUE
        raise SystemCallError.new("CreateFileW", FFI.errno)
      end

      target = "\\??\\".encode('UTF-16LE') + from_path

      rdb = REPARSE_JDATA_BUFFER.new
      rdb[:ReparseTag] = 2684354563 # IO_REPARSE_TAG_MOUNT_POINT
      rdb[:ReparseDataLength] = target.bytesize + 12
      rdb[:Reserved] = 0
      rdb[:SubstituteNameOffset] = 0
      rdb[:SubstituteNameLength] = target.bytesize
      rdb[:PrintNameOffset] = target.bytesize + 2
      rdb[:PrintNameLength] = 0
      rdb[:PathBuffer] = target

      bytes = FFI::MemoryPointer.new(:ulong)

      begin
        bool = DeviceIoControl(
          handle,
          CTL_CODE(FILE_DEVICE_FILE_SYSTEM, 41, METHOD_BUFFERED, 0),
          rdb,
          rdb[:ReparseDataLength] + rdb.header_size,
          nil,
          0,
          bytes,
          nil
        )

        error = FFI.errno

        unless bool
          RemoveDirectoryW(to_path)
          raise SystemCallError.new("DeviceIoControl", error)
        end
      ensure
        CloseHandle(handle)
      end
    end

    self
  end

  # Returns the path that a given junction points to. Raises an
  # Errno::ENOENT error if the given path does not exist. Returns false
  # if it is not a junction.
  #
  # Example:
  #
  #    Dir.mkdir('C:/from')
  #    Dir.create_junction('C:/to', 'C:/from')
  #    Dir.read_junction("c:/to") # => "c:/from"
  #
  def self.read_junction(junction)
    return false unless Dir.junction?(junction)

    junction = string_check(junction).wincode

    junction_path = (0.chr * 1024).encode('UTF-16LE')

    length = GetFullPathNameW(junction, junction_path.size, junction_path, nil)

    if length == 0 || length > junction_path.size
      raise SystemCallError.new("GetFullPathNameW", FFI.errno)
    else
      junction_path.strip!
    end

    begin
      # Generic read & write + open existing + reparse point & backup semantics
      handle = CreateFileW(
        junction_path,
        GENERIC_READ | GENERIC_WRITE,
        0,
        nil,
        OPEN_EXISTING,
        FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS,
        0
      )

      if handle == INVALID_HANDLE_VALUE
        raise SystemCallError.new("CreateFileW", FFI.errno)
      end

      rdb = REPARSE_JDATA_BUFFER.new
      rdb[:ReparseTag] = 0
      rdb[:ReparseDataLength] = 0
      rdb[:Reserved] = 0
      rdb[:SubstituteNameOffset] = 0
      rdb[:SubstituteNameLength] = 0
      rdb[:PrintNameOffset] = 0
      rdb[:PrintNameLength] = 0
      rdb[:PathBuffer] = ''

      bytes = FFI::MemoryPointer.new(:ulong)

      begin
        bool = DeviceIoControl(
          handle,
          CTL_CODE(FILE_DEVICE_FILE_SYSTEM, 42, METHOD_BUFFERED, 0),
          nil,
          0,
          rdb,
          1024,
          bytes,
          nil
        )

        error = FFI.errno

        unless bool
          raise SystemCallError.new("DeviceIoControl", error)
        end
      ensure
        CloseHandle(handle)
      end
    end

    # MSDN says print and substitute names can be in any order
    jname = (rdb[:PathBuffer].to_ptr + rdb[:SubstituteNameOffset]).read_string(rdb[:SubstituteNameLength])
    jname = jname.bytes.to_a.pack('C*')
    jname = jname.force_encoding("UTF-16LE")
    raise "Invalid junction name: #{jname.encode('UTF-8')}" unless jname[0..3] == "\\??\\".encode("UTF-16LE")
    begin
      File.expand_path(jname[4..-1].encode(Encoding.default_external))
    rescue Encoding::UndefinedConversionError
      File.expand_path(jname[4..-1].encode('UTF-8'))
    end
  end

  # Returns whether or not +path+ is empty.  Returns false if +path+ is not
  # a directory, or contains any files other than '.' or '..'.
  #
  def self.empty?(path)
    PathIsDirectoryEmptyW("#{path}".wincode)
  end

  # Returns whether or not +path+ is a junction.
  #
  def self.junction?(path)
    string_check(path)
    bool = true

    attrib = GetFileAttributesW("#{path}".wincode)

    # Only directories with a reparse point attribute can be junctions
    if attrib == INVALID_FILE_ATTRIBUTES ||
       attrib & FILE_ATTRIBUTE_DIRECTORY == 0 ||
       attrib & FILE_ATTRIBUTE_REPARSE_POINT == 0
    then
      bool = false
    end

    bool
  end

  # Class level aliases
  #
  class << self
    alias reparse_dir? junction?
  end

  private

  class << self
    # Simulate MRI's contortions for a stringiness check.
    def string_check(arg)
      return arg if arg.is_a?(String)
      return arg.send(:to_str) if arg.respond_to?(:to_str, true) # MRI honors it, even if private
      return arg.to_path if arg.respond_to?(:to_path)
      raise TypeError
    end
  end

  # Macro from Windows header file, used by the create_junction method.
  def self.CTL_CODE(device, function, method, access)
    ((device) << 16) | ((access) << 14) | ((function) << 2) | (method)
  end
end
